File: Solutions/Companies/Microsoft/WordSearchII.java
package Companies.Microsoft;

import java.util.ArrayList;
import java.util.List;

public class WordSearchII {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> l = new ArrayList<>();
        TrieNode root = createTrie(words);
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs(l, board, root, i, j);
            }
        }
        return l;
    }

    private void dfs(List<String> l, char[][] board, TrieNode root, int i, int j) {
        char c = board[i][j];
        if (c == '$' || root.children[c-'a'] == null) {
            return;
        }
        TrieNode child = root.children[c-'a'];
        if (child.word != null) {
            l.add(child.word);
            child.word = null;
        }
        board[i][j] = '$';
        if (i > 0) dfs(l, board, child, i-1, j);
        if (j > 0) dfs(l, board, child, i, j-1);
        if (i < board.length-1) dfs(l, board, child, i+1, j);
        if (j < board[0].length-1) dfs(l, board, child, i, j+1);
        board[i][j] = c;
    }

    private TrieNode createTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String w : words) {
            TrieNode cur = root;
            for (char c : w.toCharArray()) {
                if (cur.children[c-'a'] == null) {
                    cur.children[c-'a'] = new TrieNode();
                }
                cur = cur.children[c-'a'];
            }
            cur.word = w;
        }
        return root;
    }

    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        String word;
    }
}

File: Companies/Bloomberg/WordSearchII.cpp
package Companies.Bloomberg;

import java.util.ArrayList;
import java.util.List;

/**
 * find all words in the board.
 */
public class WordSearchII {

public:
    vector<string> findWords(vector<vector<char>>& board,
                             const vector<string>& words) {
        vector<string> result;

        if (board.empty() || board[0].empty() || words.empty()) {
            return result;
        }

        unique_ptr<TrieNode> root = createTrie(words);

        for (int row = 0; row < static_cast<int>(board.size()); ++row) {
            for (int col = 0;
                 col < static_cast<int>(board[0].size());
                 ++col) {
                dfs(result, board, root.get(), row, col);
            }
        }

        return result;
    }

private:
    struct TrieNode {
        array<unique_ptr<TrieNode>, 26> children{};
        string word;
    };

    const int dirs_[4][2] = {
        {0, 1},
        {0, -1},
        {-1, 0},
        {1, 0}
    };

    unique_ptr<TrieNode> createTrie(const vector<string>& words) {
        auto root = make_unique<TrieNode>();

        for (const string& word : words) {
            TrieNode* current = root.get();

            for (char c : word) {
                int index = c - 'a';

                if (!current->children[index]) {
                    current->children[index] = make_unique<TrieNode>();
                }

                current = current->children[index].get();
            }

            current->word = word;
        }

        return root;
    }

    void dfs(vector<string>& result,
             vector<vector<char>>& board,
             TrieNode* node,
             int row,
             int col) {
        char c = board[row][col];

        if (c == '#') {
            return;
        }

        int index = c - 'a';

        if (!node->children[index]) {
            return;
        }

        TrieNode* child = node->children[index].get();

        if (!child->word.empty()) {
            result.push_back(child->word);

            // 防止同一个单词通过不同路径被重复加入结果。
            child->word.clear();
        }

        board[row][col] = '#';

        for (const auto& direction : dirs_) {
            int nextRow = row + direction[0];
            int nextCol = col + direction[1];

            if (nextRow >= 0 &&
                nextRow < static_cast<int>(board.size()) &&
                nextCol >= 0 &&
                nextCol < static_cast<int>(board[0].size())) {
                dfs(result, board, child, nextRow, nextCol);
            }
        }

        board[row][col] = c;
    }

}
